### PR TITLE
Add Transformer Model

### PR DESCRIPTION
**Pull request by StoneLin0708**

### CODE REVIEW

File: lmp/model/_transformer.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
+import math
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Missing module level docstring.
Please DO follow conventions in this module (you can see from `lmp.model._lstm_model`).

File: lmp/model/_transformer.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+import torch
+import numpy as np
+from torch.utils.tensorboard import SummaryWriter
+from torch.nn import Linear
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: use `torch.nn.Linear` instead.

File: lmp/model/_transformer.py (hunk 1 above)
Comment (ProFatXuanAll) [reply lmp/model/_transformer.py]: Missing `from future import ...` backward support.

File: lmp/model/_transformer.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+from torch.nn.functional import softmax, relu
+
+
+class PositionalEncoding(torch.nn.Module):
+    "Implement the PE function."
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Always use `r"""` and `"""` to enclose docstring.
Specify which formula you use to implement positional encoding in docstring.

Comment (ProFatXuanAll) [reply lmp/model/_transformer.py]: Add type check and value check.
Do follow google python style guide.
reference: https://google.github.io/styleguide/pyguide.html

File: lmp/model/_transformer.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+    def __init__(self, d_model, dropout, max_len=5000):
+        super(PositionalEncoding, self).__init__()
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: use `super().__init__()` instead.

File: lmp/model/_transformer.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+import numpy as np
+from torch.utils.tensorboard import SummaryWriter
+from torch.nn import Linear
+from torch.nn.functional import softmax, relu
+
+
+class PositionalEncoding(torch.nn.Module):
+    "Implement the PE function."
+
+    def __init__(self, d_model, dropout, max_len=5000):
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: remove default argument.

File: lmp/model/_transformer.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        position = torch.arange(0, max_len).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2) *
+                             -(math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        x = x + torch.autograd.Variable(self.pe[:, :x.size(-2)],
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: remove old version autograd code (no need to use `Variable`, simply use `self.pe[:, :x.size(-2)]`).

File: lmp/model/_transformer.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+                                        requires_grad=False)
+        return self.dropout(x)
+
+
+class Attention(torch.nn.Module):
+    """
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: missing `r` in docstring wrapper

File: lmp/model/_transformer.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    Attention
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: MEANINGLESS comment, please DO document your code.

File: lmp/model/_transformer.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    (B, L, E) (B, L, E) (B, L, E) -> (B, L, D)
+    """
+
+    def __init__(self, E, D):
+        super().__init__()
+        self.Wq = Linear(E, D)
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Use `torch.nn.Linear` instead, do the same for the rest.

File: lmp/model/_transformer.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        self.Wk = Linear(E, D)
+        self.Wv = Linear(E, D)
+        self.D = D
+
+    def forward(self, Q, K, V, mask):
+        Q = self.Wq(Q)
+        K = self.Wk(K)
+        V = self.Wv(V)
+        Z = softmax(
+            (Q @ K.transpose(-2, -1) / math.sqrt(self.D)).masked_fill(mask == 0, -1e9), dim=-1) @ V
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Unwrap your code, do not fit all operations into just one line.

File: lmp/model/_transformer.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+        return Z
+
+
+class MultiHeadAttention(torch.nn.Module):
+    """
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Missing `r` in docstring wrapper.

File: lmp/model/_transformer.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+
+    def forward(self, Q, K, V, mask):
+        # (B, L, E) -> (B, L, D)
+        # (B, L, E) -> (B, L, H * D) * Wo(H * D, D) -> (B, L, D)
+        Z = torch.cat([att(Q, K, V, mask) for att in self.attentions], dim=-1)
+        return self.dropout(self.Wo(Z))
+
+
+class FF(torch.nn.Module):
+    """
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Missing `r` in docstring wrapper.

File: lmp/model/_transformer.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        self.W1 = Linear(E, D)
+        self.W2 = Linear(D, E)
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def forward(self, x):
+        return self.dropout(self.W2(self.dropout(relu(self.W1(x)))))
+
+
+class AddNorm(torch.nn.Module):
+    """
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Missing `r` in docstring wrapper.

File: lmp/model/_transformer.py [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+    def __init__(self, E):
+        super().__init__()
+        self.norm = torch.nn.LayerNorm(E)
+
+    def forward(self, x, sub):
+        return self.norm(x + sub)
+
+
+class DecoderLayer(torch.nn.Module):
+    """
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Missing `r` in docstring wrapper.

File: lmp/model/_transformer.py (hunk 1 above)
Comment (ProFatXuanAll) [reply lmp/model/_transformer.py]: Rename this file to `_transformer_model.py`

File: lmp/model/_transformer.py [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def forward(self, Q, K, V, mask):
+        # (B, L, E) -> (B, L, D)
+        # (B, L, E) -> (B, L, H * D) * Wo(H * D, D) -> (B, L, D)
+        Z = torch.cat([att(Q, K, V, mask) for att in self.attentions], dim=-1)
+        return self.dropout(self.Wo(Z))
+
+
+class FF(torch.nn.Module):
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Rename this class with `FeedForward`.

File: lmp/model/_transformer.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+        Z = softmax(
+            (Q @ K.transpose(-2, -1) / math.sqrt(self.D)).masked_fill(mask == 0, -1e9), dim=-1) @ V
+        return Z
+
+
+class MultiHeadAttention(torch.nn.Module):
+    """
+    Multi Head Attention:
+
+    concatenate multiple attention and reduce results by linear transform
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Write every comment as a complete sentence. (You need to capitalize every sentence and put colon.

File: lmp/model/_transformer.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+    (B, L, E) (B, L, E) (B, L, E) -> (B, L, E / H) * H -> (B, L, E)
+    """
+
+    def __init__(self, E, H, dropout):
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Do not use capital letter as variable.
Run `pylint lmp/model/_transformer.py` to check youe code.

File: lmp/model/_transformer.py (hunk 1 above)
Comment (ProFatXuanAll) [reply lmp/model/_transformer.py]: Remember to cite paper "Attention is all you need" in module level docstring.

File: lmp/model/_transformer.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+        super().__init__()
+        self.attentions = torch.nn.ModuleList(
+            [Attention(E, E // H) for _ in range(H)])
+        self.Wo = Linear(E, E)
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def forward(self, Q, K, V, mask):
+        # (B, L, E) -> (B, L, D)
+        # (B, L, E) -> (B, L, H * D) * Wo(H * D, D) -> (B, L, D)
+        Z = torch.cat([att(Q, K, V, mask) for att in self.attentions], dim=-1)
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Try to not use `torch.cat` and use `reshape` instead.

File: lmp/model/_transformer.py [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+
+    def forward(self, src):
+        L = src.shape[1]
+        src_mask = (src != self.pad_id).unsqueeze(-2)
+        subseq_mask = (torch.from_numpy(
+            np.triu(np.ones((1, L, L), dtype=np.uint8), k=1)) == 0).to(src.device)
+        return src_mask & subseq_mask
+
+
+class TransformerLanguageModel(torch.nn.Module):
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: Rename this class to `TransformerModel`.

File: lmp/model/_transformer.py [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+
+
+class SubsequentMask(torch.nn.Module):
+    def __init__(self, pad_id):
+        super().__init__()
+        self.pad_id = pad_id
+
+    def forward(self, src):
+        L = src.shape[1]
+        src_mask = (src != self.pad_id).unsqueeze(-2)
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: use `.unsueeze(0)` instead.

File: lmp/model/_transformer.py [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+        return x
+
+
+class SubsequentMask(torch.nn.Module):
+    def __init__(self, pad_id):
+        super().__init__()
+        self.pad_id = pad_id
+
+    def forward(self, src):
+        L = src.shape[1]
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: use `src.size(-1)` instead

File: lmp/model/_transformer.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+        src_mask = (src != self.pad_id).unsqueeze(-2)
+        subseq_mask = (torch.from_numpy(
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: use `torch.tril` instead.

File: lmp/model/_transformer.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+        num_linear_layers: int,
+        num_rnn_layers: int,
+        pad_token_id: int,
+        vocab_size: int
+    ):
+        super().__init__()
+
+        self.subseqmask = SubsequentMask(pad_id=pad_token_id)
+        self.pad_id = pad_token_id
+        self.pad_id = pad_token_id
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: duplicate code

Comment (ProFatXuanAll) [reply lmp/model/_transformer.py]: no need to save `pad_token_id` as attribute (remove `self.pad_id`).

File: lmp/model/_transformer.py [hunk 24]
```diff
@@ -0,0 +1,199 @@
...
+    def __init__(self, E, H, Dff, dropout):
+        super().__init__()
+        self.att = MultiHeadAttention(E=E, H=H, dropout=dropout)
+        self.addnorm1 = AddNorm(E=E)
+        self.addnorm2 = AddNorm(E=E)
+        self.ff = FF(E=E, D=Dff, dropout=dropout)
+        self.dropout = torch.nn.Dropout(dropout)
+
+    def forward(self, x, mask):
+        x = self.addnorm1(x, self.att(x, x, x, mask))
```
Comment (ProFatXuanAll) [inline lmp/model/_transformer.py]: missing dropout layer
